### PR TITLE
Add Adaptive Sampler

### PR DESCRIPTION
**Pull request by black-adder**
- [x] Tests

This PR doesn't include initializing and updating the adaptive sampler. I'll do that in a subsequent PR.

### CODE REVIEW

File: jaeger_client/sampler.py [hunk 1]
```diff
@@ -177,6 +178,100 @@ def __str__(self):
...
+    """
+    def __init__(self, operation, lower_bound, rate):
+        super(GuaranteedThroughputProbabilisticSampler, self).__init__(
+            tags={
+                SAMPLER_TYPE_TAG_KEY: SAMPLER_TYPE_LOWER_BOUND,
+                SAMPLER_PARAM_TAG_KEY: rate,
+            }
+        )
+        self.probabilistic_sampler = ProbabilisticSampler(rate)
+        self.lower_bound_sampler = RateLimitingSampler(lower_bound)
```
Comment (yurishkuro) [inline jaeger_client/sampler.py]: please double-check that this class works with values `<1`, as at some point the very first implementation (not sure in which language) only worked with qps expressed as a whole number

Comment (black-adder) [inline jaeger_client/sampler.py]: I checked the code and it looks like it works. I added a test to make sure.

File: tests/test_sampler.py [hunk 2]
```diff
@@ -116,13 +123,72 @@ def test_rate_limiting_sampler():
...
+
+def test_adaptive_sampler():
+    sampling_rates = [
+        OperationSamplingStrategy('op', ProbabilisticSamplingStrategy(0.5))
+    ]
+    strategies = PerOperationSamplingStrategies(0.51, 3, sampling_rates)
+
+    sampler = AdaptiveSampler(strategies, 2)
+    id1 = 1L << 63  # second most significant bit, dividing full range in half
+    assert id1 == 0x8000000000000000L
```
Comment (yurishkuro) [inline tests/test_sampler.py]: we keep doing this, let's define  constant once and have a standalone test that validates that the constant is equal to `1L << 63` expression

File: tests/test_sampler.py [hunk 3]
```diff
@@ -116,13 +123,72 @@ def test_rate_limiting_sampler():
...
+    # This operation is seen for the first time by the sampler
+    sampled, tags = sampler.is_sampled(id1-10, "new_op")
+    assert sampled
+    assert tags == get_tags('probabilistic', 0.51)
+
+    # This operation is seen for the first time by the sampler but surpasses
+    # max_operations of 2
+    sampled, tags = sampler.is_sampled(id1-10, "new_op_2")
+    assert sampled
+    assert tags == get_tags('probabilistic', 0.51)
```
Comment (yurishkuro) [inline tests/test_sampler.py]: so there's no difference with new_op. Is there a way we can distinguish these two scenarios by asserting something else?

**Pull request merged**